### ISSUE TITLE
Reverse transform

### ISSUE THREAD

**Issue by stity** [fepegar/torchio#299]
**🚀 Feature**
Reverse the history of transforms applied to a subject.

**Motivation**

It would be great able to reverse the transforms of a subject especially for spatial transforms.
The idea is to use test-time augmentation (TTA) in order to have a better measure of uncertainty and have more robust predictions.
For reference : https://openreview.net/pdf?id=Byxv9aioz 

**Pitch**
Add a `reverse_transform`  method to the subject class or any other equivalent.
This methode takes a subject or an image as argument, just like transforms.

```python
pipeline = Compose([
    RandomAffine(),
    RandomElasticDeformation()
])
def get_TTA_segmentation:
    augmented_subject = pipeline(subject)
    segmentation = classify_subject(augmented_subject) # segmentation given by the neural network in the augmented subject space
    augmented_subject.reverse_transform(segmentation) # segmentation in the original subject space
```

Obviously some transform such as crop introduce a loss of information and are therefore not 'invertible'.
These case should be carefully handled : should `reverse_transform` of `Crop` raise an error or apply an equivalent `Pad` transform ?

**Alternatives**

No alternative considered.

**Additional context**

**Comment by romainVala** [comment]
Sorry for the confusion, but I post a bug (in the wrong issue) I should have make it here, since it is related to inverted rotation :
https://github.com/fepegar/torchio/issues/336#issuecomment-718481336
can you answer here ...

**Comment by fepegar** [comment]
@romainVala 

> By the way, well done for the new plotting, it is indeed convenient ! (note a small bug If I remove both 'brain' and 'head' images it won't plot if only "t1"...)

I'm glad it's useful! Thanks for reporting that, I fixed it in 6e95bb7.

---

The second bug is weirder. I'm investigating it. I do really want to be able to invert spatial transforms easily, for things like test-time augmentation or contrastive learning.

**Comment by romainVala** [comment]
yes it is an interesting feature, from my understanding, it is somewhat related to different discussions we had on reproducing a given transform. Since to be reversed, one should be able from the history to get all needed parameter
we have a working solution (on my fork) which is (almost) proposed here
https://github.com/fepegar/torchio/pull/287
I think we should conclude there before starting this one, or do you have something else in mind ?

**Comment by fepegar** [comment]
Yeah I need to get back to that PR!

**Comment by fepegar** [comment]
I think I got it. It's an issue about `sitk_image.GetSize()` and `sitk.GetArrayFromImage(sitk_image)` being reversed. The image center is not being computed correctly.

**Comment by fepegar** [comment]
Ok, I think I know what's going on. It _only_ took my about 3 hours.

Let's say we want to rotate an image θ degrees around the x and y axes. The rotation matrix is Rxy = Ry * Rx. If we then want to apply a rotation of -θ degrees, the rotation matrix will be Rxy2 = Ry' * Rx', where ' denotes the inverse transform.

The problem is that Rxy != Rxy2. What we need is Rxy' = Rx' * Ry'.

So it's not as simple as just using the original angle times -1, there really should be a way of passing an affine matrix and let SimpleITK compute the inverse transform appropriately.

**Comment by romainVala** [comment]
Yes you are right, the combination of simple rotation translation and scaling are not commutative, so the order is important (and is an arbitrary choice, which may be should be documented ...)  and indeed what matter at the end is the resulting affine. I currently store the single parameter (rot trans scale) but indeed one should store the affine (for inversion purpose)

**Comment by fepegar** [comment]
What do you think of this pseudocode?

```python
transform = RandomTransform(settings)

transformed = transform(image)
params = transformed.params

same_transform = Transform(*params.args, **params.kwargs)
inv_transform = Transform(*params.args, **params.kwargs, invert=True)

same = same_transform(image)
original = inv_transform(transformed)
```

The idea would be to have a functional API with classes that take as args and kwargs the parameters generated by their random counterparts. Maybe this could help us get rid of relying on that seed to reproduce results.

**Comment by fepegar** [comment]
> Let's say we want to rotate an image θ degrees around the x and y axes. The rotation matrix is Rxy = Ry * Rx. If we then want to apply a rotation of -θ degrees, the rotation matrix will be Rxy2 = Ry' * Rx', where ' denotes the inverse transform.
The problem is that Rxy2 != Rxy'. What we need is Rxy' = Rx' * Ry'.

For reference, I believe ITK actually computes R as Rx * Ry * Rz, not Ry * Ry * Rx.

**Comment by romainVala** [comment]
Thanks, for the reference. It would be nice, if you can get the complete expression which include scaling and translation too (useful for inter-software comparison ... )

About the inverse transform, may be an easier way  is to add an argument called inverse, which would do the exact same affine computation , but take the inverse of the affine, just before to apply it, 
Doing so we could keep working with more intuitive parameter (scalling rotation translation, and getting them from the history), and just adding inverse=True will do the job !

**Comment by romainVala** [comment]
> What do you think of this pseudocode?

oups I did not see this before, my answer, but it seems we do agree ...
I do like your proposition, although I do not understand why you need 
*params.args, **params.kwargs

what differences in the content ?

I like it because it is explicit and you can get the exact parameters used. (indeed you can also have a need of the parameters for further use (comparison of model error as a function of rotation angle in x direction for instance)

Nevertheless the seeding mechanism may still be need, for certain transform: RandomNoise will have one random parameter: std which you can store and get, but is not enough to reproduce the exact same transform without the same seeding (which is an important feature)
by the way what would be the inverse here ? 
reproducing the 3D noise and substract it voxel bu voxel ? 
what the use of it ?  (I need to read the reference in more detail ...)

**Comment by romainVala** [comment]
For elasticDeformation, the seeding will avoid to store the all displacement field, which can be quite big

**Comment by romainVala** [comment]
Thanks @stity  for the reference, I read in more detail the article, and I found it very interesting (this is exactly what I am trying to do, but I was more concentrate on the transformation-based uncertainty for a given model, than the test time prediction). There are some technical problem that are not discuss in the article (may be it is not the place here but I give a short try)
1_  applying the inverse transform to the label, will modify the label (interpolation issue) so this will add extra variance (independently of model prediction, and input data). How much (this is why I open the issue https://github.com/fepegar/torchio/issues/342, to reduce this extra noise)
2_ which parameter range for the transformed:  they give an example for fetus segmentation where any spatial position is expected, so ok, but what about noise ? adding more and more noise will increase the prediction variance. So the estimated uncertainty will be directly correlated with the maximum std for noise ...

**Comment by romainVala** [comment]
Before reading the article, I understood inverse transform in a general sens and to be applied to the image, 
(so inverse a random bias field is to applied an inversed multiplication field which will led to the original image)
But from a test time perspective (if I understand it correctly) the inverse is only for the label space. In this case the inverse of all intensity transformed is identity right ?

So about @fepegar about the pseudo code you propose, it may be restriction to spatial transforms ... ?

**Comment by fepegar** [comment]
Yes, I was mostly thinking of adding this functionality to spatial transforms.

**Comment by stity** [comment]
Indeed, only spatial transforms need to be inversed.

@romainVala 
For the interpolation, I suspect that they are using linear interpolation on the prediction map since it is a floating point map.

For the amount of augmentation, I think it should stay realistic and/or match the augmentation done on the training set. Even though it is expected that the variance of the prediction increase with the amount of augmentation, you should be more interested in the relative values (voxel to voxel) rather than the absolute values of the variance (between two different experiments with different levels of TTA).

**Comment by romainVala** [comment]
good point for the interpolation, I guess it make sense to interpolation the prediction map, before to make it binary

about relative values (voxel to voxel) ... I do not understand, 
I though we just  want to compute the variance (between different levels of TTA) ...

**Comment by fepegar** [comment]
> For the interpolation, I suspect that they are using linear interpolation on the prediction map since it is a floating point map.

If you have any specific question about that paper, I can ask. It's from my lab!

**Comment by romainVala** [comment]
About inverting the Elastic deformation fernando propose : https://github.com/fepegar/torchio/issues/342#issuecomment-721881564

Your solution is no more working because of the recent PR merge https://github.com/fepegar/torchio/commit/8b6296402fc694e25d9a7de46b754c84bd2c2bbe
the 'coarse_grid' is no more store in the history, (I try to add it by adding the line 
`self.coarse_grid = bspline_params`
but the current history store (in string) does not work for big matrix (This need to be discuss else where if needed) but I find a solution without the need to store the big coarse_grid matrix.

before to propose a PR, I explain here to see if it seems ok
I add the input parameter get_inverse (default False) in  radom_elastic_deformation.py AND just after the get_param() call in apply_transform I add :
```python
        if self.get_inverse:
            bspline_params = -bspline_params
```

Then now to get the inverse transform it is as simple as  (thanks @GReguig , for the seeding mechanisme)

```python
import torchio as tio

colin = tio.datasets.Colin27()
transform = tio.RandomElasticDeformation()
transformed = transform(colin)

trsfm_hist, seeds_hist = tio.compose_from_history(history=transformed.history)
trsfm_hist[0].get_inverse = True
colin_back = trsfm_hist[0](transformed, seed=seeds_hist[0])

```

it should be easy to make the same for random_affine

**Comment by fepegar** [comment]
Hi, @romainVala. I suppose that would work, but it seems a bit hacky. I will open a new issue to discuss further development on randomness and inversion of transforms. I have some ideas and I think you will like them.

I'm not sure why the coarse grid cannot be saved. The array is quite small. Also, saving the parameters as strings might not be ideal...

With the PR that I will open soon:

```python
import sys
import torchio as tio

name = sys.argv[1].capitalize()
tname = f'Random{name}'

colin = tio.datasets.Colin27()
transform = getattr(tio, tname)()
print('Random transform:', transform)  # RandomAffine
transformed = transform(colin)
h = transformed.history[0]
transform_name, arguments = transformed.history[0]
same_transform = getattr(tio, transform_name)(**arguments)  # TODO: nice wrapper for this
print('Transform to reproduce:', same_transform)  # Affine

transformed_reproduced = same_transform(colin)
print('Diff:', (transformed_reproduced.t1.numpy() - transformed.t1.numpy()).sum())  # 0.0

transformed.t1.save(f'/tmp/colin_{name}.nii.gz')
transformed_reproduced.t1.save(f'/tmp/colin_{name}_same.nii.gz')

if same_transform.is_invertible():
    transform_inversed = same_transform.inverse()
    print('Inversed transform:', transform_inversed)  # Affine (new one)
    colin_back = transform_inversed(transformed)
    # Try to avoid padded values
    a = colin_back.t1.numpy()[0, 80:100, 80:100, 80:100]
    b = colin.t1.numpy()[0, 80:100, 80:100, 80:100]
    print('Diff:', (a - b).mean())  # value depends on initial range, but ideally 0.0
    colin_back.t1.save(f'/tmp/colin_{name}_inverse.nii.gz')
```
